//! Error handling related code

use axum::response::IntoResponse;
use http::StatusCode;

// Make our own error that wraps `anyhow::Error`.
pub struct AppError(pub anyhow::Error);

// Tell axum how to convert `AppError` into a response.
impl IntoResponse for AppError {
    fn into_response(self) -> axum::response::Response {
        let msg = self.0.to_string();
        tracing::error!("{msg}");
        (
            StatusCode::INTERNAL_SERVER_ERROR,
            format!("Something went wrong: {msg}"),
        )
            .into_response()
    }
}

// This enables using `?` on functions that return `Result<_, anyhow::Error>` to turn them into
// `Result<_, AppError>`. That way you don't need to do that manually.
impl<E> From<E> for AppError
where
    E: Into<anyhow::Error>,
{
    fn from(err: E) -> Self {
        Self(err.into())
    }
}
